core/lib/dal/src/events_web3_dal.rs

```diff
@@ -118,18 +128,69 @@ impl EventsWeb3Dal<'_, '_> {
 
         where_sql += &format!(" AND (miniblock_number <= {})", filter.to_block.0 as i64);
 
-        if !filter.addresses.is_empty() {
-            where_sql += &format!(" AND (address = ANY(${}))", arg_index);
+        // Add filters for address (like `address = ANY($1)` or `address = $1`)
+        if let Some(filter_sql) =
+            Self::build_sql_filter(filter.addresses.len() as u32, "address", arg_index)
+        {
+            where_sql += &filter_sql;
             arg_index += 1;
         }
-        for (topic_index, _) in filter.topics.iter() {
-            where_sql += &format!(" AND (topic{} = ANY(${}))", topic_index, arg_index);
-            arg_index += 1;
+
+        // Add filters for topics (like `topic0 = ANY($2)`)
+        for (topic_index, topics) in filter.topics.iter() {
+            if let Some(filter_sql) = Self::build_sql_filter(
+                topics.len() as u32,
+                &format!("topic{}", topic_index),
+                arg_index,
+            ) {
+                where_sql += &filter_sql;
+                arg_index += 1;
+            }
         }
 
         (where_sql, arg_index)
     }
 
+    // Builds SQL filter for optional filter (like address or topics).
+    fn build_sql_filter(
+        number_of_entities: u32,
+        field_name: &str,
+        arg_index: u8,
+    ) -> Option<String> {
+        match number_of_entities {
+            0 => None,
+            1 => Some(format!(" AND ({} = ${})", field_name, arg_index)),
+            _ => Some(format!(" AND ({} = ANY(${}))", field_name, arg_index)),
+        }
+    }
+
+    // Binds parameters for optional filter (like address or topics).
+    // Noop if there are no values.
+    // Assumes `=$1` syntax for single value and `=ANY($1)` for multiple values.
+    // See the method above for details.
+    fn bind_params_for_optional_filter_query_as<'q, O>(
+        query: QueryAs<'q, Postgres, O, PgArguments>,
+        values: Vec<&'q [u8]>,
+    ) -> QueryAs<'q, Postgres, O, PgArguments> {
+        match values.len() {
+            0 => query,
+            1 => query.bind(values[0]),
+            _ => query.bind(values),
+        }
+    }
+
+    // Same as `bind_params_for_optional_filter_query_as` but for `Query` instead of `QueryAs`.
+    fn bind_params_for_optional_filter_query<'q>(
+        query: Query<'q, Postgres, PgArguments>,
+        values: Vec<&'q [u8]>,
+    ) -> Query<'q, Postgres, PgArguments> {
+        match values.len() {
+            0 => query,
+            1 => query.bind(values[0]),
+            _ => query.bind(values),
+        }
+    }
```

Comment:
This is not good - but I couldn't find a way to write this in a generic manner (`.bind` is defined sepraately on `Query` and `QueryAs`. `QueryAs`  has `inner` of type `Query`, but it's `pub(crate)`